### PR TITLE
[#4072] Added Catch2 unit-testing framework and one test.

### PR DESCRIPTION
**Pull request by korydraughn**
Depends on: [irods/externals#14]

Passed CI.

### CODE REVIEW

**Comment by korydraughn** [comment]
Should the folder name be adjusted?  Someone may think the library is labeled **Catch** instead of **Catch2**.

image

**Comment by trel** [comment]
folder name is just constructed from project name and version - and this project name ends in an integer.

looks strange, but should be fine.

File: unit_tests/test_irods_linked_list_iterator.cpp [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+    const int list_length = 5;
+    auto list = make_c_linked_list(list_length);
+    irods::linked_list_iterator<dataObjInfo_t> pos{list.get()};
+
+    REQUIRE(list.get() != nullptr);
+    REQUIRE(list.get() == &*pos);
+
+    SECTION("copy_assn")
+    {
+        auto copy = pos;
```
Comment (tempoz) [inline unit_tests/test_irods_linked_list_iterator.cpp]: const

File: unit_tests/test_irods_linked_list_iterator.cpp [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+        REQUIRE(copy == pos);
+    }
+
+    SECTION("pre_fix_inc")
+    {
+        auto copy = ++pos;
```
Comment (tempoz) [inline unit_tests/test_irods_linked_list_iterator.cpp]: const

File: unit_tests/test_irods_linked_list_iterator.cpp [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+        REQUIRE(copy == pos);
+    }
+
+    SECTION("post_fix_inc")
+    {
+        auto previous_pos = pos++;
```
Comment (tempoz) [inline unit_tests/test_irods_linked_list_iterator.cpp]: const

File: unit_tests/test_irods_linked_list_iterator.cpp [hunk 4]
```diff
@@ -0,0 +1,90 @@
...
+    SECTION("std_algos")
+    {
+        REQUIRE(std::distance(begin(list.get()), end(list.get())) == list_length);
+    }
+
+    SECTION("range_based_for_loop")
+    {
+        int count = 0;
+
+        for (auto _ : list.get())
```
Comment (tempoz) [inline unit_tests/test_irods_linked_list_iterator.cpp]: const

File: CMakeLists.txt [hunk 5]
```diff
@@ -1310,6 +1310,7 @@ add_subdirectory(plugins/network)
 add_subdirectory(plugins/resources)
 add_subdirectory(plugins/rule_engines)
 add_subdirectory(test/c_api_test)
+add_subdirectory(unit_tests)
```
Comment (alanking) [inline CMakeLists.txt]: Thoughts on putting this under the test directory? Or do we want to phase that out eventually?

**Pull request merged**